core/sys/isrControl.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2015, Scanimetrics - http://www.scanimetrics.com
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/*---------------------------------------------------------------------------*/
+#ifndef INTERRUPTS_H_
+#define INTERRUPTS_H_
+/*---------------------------------------------------------------------------*/
+/**
+* Includes
+**/
+#include "cc.h"
+#include "isrControlArch.h"
+/*---------------------------------------------------------------------------*/
+/**
+ * Disables global interrupts, returns true if interrupts were already disabled
+**/
+static CC_INLINE int _arch_disableINTS();
+/**
+ * Enables global interrupts
+**/
+static CC_INLINE void _arch_enableINTS();
+/*---------------------------------------------------------------------------*/
+/**
+* Macros
+**/
+#define disableINTS _arch_disableINTS
+#define enableINTS  _arch_enableINTS
+
+#define ATOMIC_SECTION_ENTER \
+    {int _ints_disabled = _arch_disableINTS();
+
+#define ATOMIC_SECTION_EXIT \
+    if(!_ints_disabled) _arch_enableINTS();}
+/*---------------------------------------------------------------------------*/
```

Comment:
@tonnenpinguin 

Honestly I agree. The compiler errors produced by this approach can indeed be very surprising and I really don't like that.

 Regarding the use of a global variable in this way, the problem with that approach is that it does not account for the possibility that ATOMIC_SECTIONs might be nested (for example if a function within an atomic section tries to create its own atomic section).

The only way I know of that both allows nested sections and avoids the weirdness with my current macros is to use a global counter (lock) variable which is incremented for each interrupt disable and decremented for each enable.

This is the approach used in the CMSIS library for energy micro (see INT_Disable and INT_Enable in https://code.google.com/p/rt-thread/source/browse/trunk/bsp/efm32/Libraries/efm32lib/inc/efm32_int.h?r=1525). The only reason I don't like this approach is because of the possibility of integer overflow (but maybe it is reasonable to just assume that this will never happen in any real software).